warn-stranded.lua

```diff
@@ -0,0 +1,581 @@
+        ' '..getSexString(unit.sex)..' Stress category: '..dfhack.units.getStressCategory(unit)
+end
+
+-- Use group data, index, and command arguments to generate a group
+--   designation string.
+local function getGroupDesignation(group, groupIndex, walkGroup)
+    local groupDesignation = ''
+
+    if group['mainGroup'] then
+        groupDesignation = ' (Main Group)'
+    else
+        groupDesignation = ' (Group '..groupIndex..')'
+    end
+
+    if walkGroup then
+        groupDesignation = groupDesignation..' {'..group.walkGroup..'}'
+    end
+
+    return groupDesignation
+end
+
+-- Add unit.id to text
+local function addId(text, unit)
+    return text..'|'..unit.id..'| '
+end
+
+-- Uses persistent API. Low-level, deserializes 'warnStrandedIgnored' key and
+--   will return an initialized empty warnStrandedIgnored table if needed.
+-- Performance characterstics unknown of persistent API
+local function deserializeIgnoredUnits()
+    local currentIgnore = dfhack.persistent.get('warnStrandedIgnore')
+    if currentIgnore == nil then return {} end
+
+    local tbl = {}
+
+    for v in string.gmatch(currentIgnore['value'], '%d+') do
+        table.insert(tbl, v)
+    end
+
+    return tbl
+end
+
+-- Uses persistent API. Deserializes 'warnStrandedIgnore' key to determine if unit is ignored
+--   deserializedIgnores is optional but allows us to only call deserialize once like an explicit cache.
+local function unitIgnored(unit, deserializedIgnores)
+    local ignores = deserializedIgnores or deserializeIgnoredUnits()
+
+    for index, id in ipairs(ignores) do
+        if tonumber(id) == unit.id then
+            return true, index
+        end
+    end
+
+    return false
+end
+
+-- Check for and potentially add [IGNORED] to text.
+--   Optional deserializedIgnores allows us to call deserialize once for a group of operations
+local function addIgnored(text, unit, deserializedIgnores)
+    if unitIgnored(unit, deserializedIgnores) then
+        return text..'[IGNORED] '
+    end
+
+    return text
+end
+
+-- Uses persistent API. Toggles a unit's ignored status by deserializing 'warnStrandedIgnore' key
+--   then serializing the resulting table after the toggle.
+-- Optional cache parameter could affect data integrity. Make sure you don't need data reloaded
+--   before using it. Calling several times in a row can use the return result of the function
+--   as input to the next call.
+local function toggleUnitIgnore(unit, deserializedIgnores)
+    local ignores = deserializedIgnores or deserializeIgnoredUnits()
+    local is_ignored, index = unitIgnored(unit, ignores)
+
+    if is_ignored then
+        table.remove(ignores, index)
+    else
+        table.insert(ignores, unit.id)
+    end
+
+    dfhack.persistent.delete('warnStrandedIgnore')
+    dfhack.persistent.save({key = 'warnStrandedIgnore', value = table.concat(ignores, ' ')})
```

Comment:
I think this might be an inefficient way of going about storing state. you can store each unit's data in its own key and use prefix matching to retrieve the full set of keys. the plugins tend to use this strategy, like `seedwatch`, `logistics`, `autobutcher`, and others. see `dfhack.persistent.get_all()` for the Lua version of the API.

You can also cache the state in a global that is reset on world load. example: https://github.com/DFHack/scripts/blob/master/prioritize.lua#L586

Reply:
Oh, awesome. Yeah this was very hacked together and I wasn't sure the best way to use the persistent API here. I'll look at those examples and get something better going.